File: java21/src/test/java/io/github/sps23/designpatterns/facade/OrderFulfillmentFacadeTest.java
package io.github.sps23.designpatterns.facade;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.math.BigDecimal;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

@DisplayName("OrderFulfillmentFacade Java 21 Tests")
class OrderFulfillmentFacadeTest {

    @Test
    @DisplayName("Should place an order through all subsystems")
    void shouldPlaceOrderThroughFacade() {
        NotificationRecorder notificationRecorder = new NotificationRecorder();
        OrderFulfillmentFacade facade = new OrderFulfillmentFacade(new InventoryAlwaysAvailable(),
                new PaymentAlwaysAccepted(), new ShippingAlwaysScheduled(), notificationRecorder);

        var request = new OrderFulfillmentFacade.OrderRequest("customer-42", "SKU-42", 2,
                new BigDecimal("149.99"), "12 Main Street, London");

        OrderFulfillmentFacade.FulfillmentResult result = facade.placeOrder(request);

        assertTrue(result.success());
        assertEquals("Order placed successfully", result.message());
        assertNotNull(result.trackingId());
        assertTrue(notificationRecorder.sent);
    }

    @Test
    @DisplayName("Should reject when inventory is insufficient")
    void shouldRejectWhenInventoryIsInsufficient() {
        OrderFulfillmentFacade facade = new OrderFulfillmentFacade(new InventoryInsufficient(),
                new PaymentAlwaysAccepted(), new ShippingAlwaysScheduled(),
                new NotificationRecorder());

        var request = new OrderFulfillmentFacade.OrderRequest("customer-99", "SKU-999", 10,
                new BigDecimal("40.00"), "99 Market Road, Berlin");

        OrderFulfillmentFacade.FulfillmentResult result = facade.placeOrder(request);

        assertFalse(result.success());
        assertEquals("Inventory unavailable for SKU SKU-999", result.message());
    }

    @Test
    @DisplayName("Should reject when payment is declined")
    void shouldRejectWhenPaymentIsDeclined() {
        OrderFulfillmentFacade facade = new OrderFulfillmentFacade(new InventoryAlwaysAvailable(),
                new PaymentRejected(), new ShippingAlwaysScheduled(), new NotificationRecorder());

        var request = new OrderFulfillmentFacade.OrderRequest("blocked-customer", "SKU-42", 1,
                new BigDecimal("10.00"), "1 High Street, Paris");

        OrderFulfillmentFacade.FulfillmentResult result = facade.placeOrder(request);

        assertFalse(result.success());
        assertEquals("Payment failed for customer blocked-customer", result.message());
    }

    private static final class InventoryAlwaysAvailable
            implements
                OrderFulfillmentFacade.InventoryGateway {
        @Override
        public boolean hasStock(String sku, int quantity) {
            return true;
        }
    }

    private static final class InventoryInsufficient
            implements
                OrderFulfillmentFacade.InventoryGateway {
        @Override
        public boolean hasStock(String sku, int quantity) {
            return false;
        }
    }

    private static final class PaymentAlwaysAccepted
            implements
                OrderFulfillmentFacade.PaymentGateway {
        @Override
        public boolean charge(String customerId, BigDecimal amount) {
            return true;
        }
    }

    private static final class PaymentRejected implements OrderFulfillmentFacade.PaymentGateway {
        @Override
        public boolean charge(String customerId, BigDecimal amount) {
            return false;
        }
    }

    private static final class ShippingAlwaysScheduled
            implements
                OrderFulfillmentFacade.ShippingGateway {
        @Override
        public String scheduleShipment(String customerId, String shippingAddress) {
            return "TRACK-123";
        }
    }

    private static final class NotificationRecorder
            implements
                OrderFulfillmentFacade.NotificationGateway {
        private boolean sent;

        @Override
        public void sendConfirmation(String customerId, String sku, String trackingId) {
            sent = true;
        }
    }
}

File: kotlin/src/test/kotlin/io/github/sps23/designpatterns/facade/OrderFulfillmentFacadeTest.kt
package io.github.sps23.designpatterns.facade

import io.github.sps23.designpatterns.facade.OrderFulfillmentFacade.OrderRequest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import java.math.BigDecimal

@DisplayName("OrderFulfillmentFacade Kotlin Tests")
class OrderFulfillmentFacadeTest {
    @Test
    fun `should place an order through all subsystems`() {
        val recorder = NotificationRecorder()
        val facade =
            OrderFulfillmentFacade(
                inventoryGateway = InventoryAlwaysAvailable(),
                paymentGateway = PaymentAlwaysAccepted(),
                shippingGateway = ShippingAlwaysScheduled(),
                notificationGateway = recorder,
            )

        val request =
            OrderRequest(
                customerId = "customer-42",
                sku = "SKU-42",
                quantity = 2,
                amount = BigDecimal("149.99"),
                shippingAddress = "12 Main Street, London",
            )

        val result = facade.placeOrder(request)

        assertTrue(result.success)
        assertEquals("Order placed successfully", result.message)
        assertNotNull(result.trackingId)
        assertTrue(recorder.sent)
    }

    @Test
    fun `should reject when inventory is insufficient`() {
        val facade =
            OrderFulfillmentFacade(
                inventoryGateway = InventoryInsufficient(),
                paymentGateway = PaymentAlwaysAccepted(),
                shippingGateway = ShippingAlwaysScheduled(),
                notificationGateway = NotificationRecorder(),
            )

        val result =
            facade.placeOrder(
                OrderRequest(
                    customerId = "customer-99",
                    sku = "SKU-999",
                    quantity = 10,
                    amount = BigDecimal("40.00"),
                    shippingAddress = "99 Market Road, Berlin",
                ),
            )

        assertFalse(result.success)
        assertEquals("Inventory unavailable for SKU SKU-999", result.message)
    }

    @Test
    fun `should reject when payment is declined`() {
        val facade =
            OrderFulfillmentFacade(
                inventoryGateway = InventoryAlwaysAvailable(),
                paymentGateway = PaymentRejected(),
                shippingGateway = ShippingAlwaysScheduled(),
                notificationGateway = NotificationRecorder(),
            )

        val result =
            facade.placeOrder(
                OrderRequest(
                    customerId = "blocked-customer",
                    sku = "SKU-42",
                    quantity = 1,
                    amount = BigDecimal("10.00"),
                    shippingAddress = "1 High Street, Paris",
                ),
            )

        assertFalse(result.success)
        assertEquals("Payment failed for customer blocked-customer", result.message)
    }

    private class InventoryAlwaysAvailable : OrderFulfillmentFacade.InventoryGateway {
        override fun hasStock(
            sku: String,
            quantity: Int,
        ): Boolean = true
    }

    private class InventoryInsufficient : OrderFulfillmentFacade.InventoryGateway {
        override fun hasStock(
            sku: String,
            quantity: Int,
        ): Boolean = false
    }

    private class PaymentAlwaysAccepted : OrderFulfillmentFacade.PaymentGateway {
        override fun charge(
            customerId: String,
            amount: BigDecimal,
        ): Boolean = true
    }

    private class PaymentRejected : OrderFulfillmentFacade.PaymentGateway {
        override fun charge(
            customerId: String,
            amount: BigDecimal,
        ): Boolean = false
    }

    private class ShippingAlwaysScheduled : OrderFulfillmentFacade.ShippingGateway {
        override fun scheduleShipment(
            customerId: String,
            shippingAddress: String,
        ): String = "TRACK-123"
    }

    private class NotificationRecorder : OrderFulfillmentFacade.NotificationGateway {
        var sent = false

        override fun sendConfirmation(
            customerId: String,
            sku: String,
            trackingId: String,
        ) {
            sent = true
        }
    }
}
